rfcs/000-caller-side-managed-buffers.md

```diff
@@ -0,0 +1,833 @@
+
+        State::NeedsSignature { message, version } => {
+            let mut opt = if let Some(handle) = handle.take() {
+                handle.await.ok()
+            } else {
+                None
+            };
+
+            let dss = dss.as_ref().ok_or(Error::RelevantErrorType)?;
+
+            let res = (|| {
+                let (cert_verifier, cert_verified, certificate_entries) =
+                    opt.take().ok_or(Error::RelevantErrorType)?;
+
+                let sig_verified = match version {
+                    ProtocolVersion::TLSv1_2 => {
+                        cert_verifier.verify_tls12_signature(&message, &end_entity.cert, dss)?
+                    }
+
+                    ProtocolVersion::TLSv1_3 => {
+                        cert_verifier.verify_tls13_signature(&message, &end_entity.cert, dss)?
+                    }
+                };
+
+                (cert_verified, sig_verified)
+            })();
+
+            let outcome = match res {
+                Ok((cert_verified, sig_verified)) => CertificateVerificationOutcome::Success {
+                    cert_verified,
+                    sig_verified,
+                }
+
+                Err(_) = CertificateVerificationOutcome::Failure,
+            };
+
+            conn.certificate_verification_outcome(outcome);
+        }
+    }
+}
+```
+
+Note that some (most?) async runtimes provide a task API with "detach on drop" semantics.
+To not waste resources, the handling of any fatal error while the async task is in flight should also cancel the task.
+
+## The `dangerous_configuration` feature
+
+The proposed API requires making types like `ServerCertVerified`, `HandshakeSignatureValid`, etc. public.
+These types are currently gated behind the `dangerous_configuration` Cargo feature.
+
+The needed types will become public regardless of the state of the `dangerous_configuration` feature but
+their constructors, for example `ServerCertVerified::assertion`, will continue to be feature gated.
+
+# Alternatives
+
+## More eager `State`
+
+Some of the `State` variants are unit structs.
+When handling those variants, one must call API like `decrypt_early_data`.
+In some cases this can feel like one is performing redundant checks:
+
+```rust
+if status.state == State::ReceivedEarlyData {
+    if let Some(early_data) = conn.decrypt_early_data(&mut incoming_tls) {
+        // ..
+
+        conn.discard_early_data();
+    }
+}
+```
+
+An alternative to this is to, for example, include the early data in the `State` variant itself.
+
+```rust
+enum State<'a> {
+    ReceivedEarlyData(&'a [u8]),
+}
+```
+
+But this alternative has two issues:
+`State`, and thus `Status` as well, would have to gain a lifetime that ties it to the `incoming_tls` buffer.
+Given how Rust works, even if `State` was not in the `ReceivedEarlyData` variant, the `status.state` field would still "freeze" `incoming_tls` meaning that it can't be mutated until `status.state` goes out of scope / is dropped.
```

Comment:
Why is this a problem?

Reply:
I was a bit concerned with end users running into borrow checker problems in the case of `State` borrowing the TLS buffers but after doing `sketch2` I think my concerns have been somewhat alleviated. I'll do some further testing with a mocked async IO API to confirm that the borrow checker is still happy with borrows across suspension (`.await`) points.